### ISSUE TITLE
[Bug - ArmCord client/arRPC only] Freeze/not start (until acted on)

### ISSUE THREAD

**Issue by TheAlienDrew** [RiiConnect24/RiiTag-RPC#29]
OS: Raspberry Pi OS (64-bit)

After starting the application (assume already logged in), it appears frozen, until you either resize the window or press an arrow key to move the selection, then it starts to work.

**Comment by DismissedGuy** [comment]
Window drawing is completely left up to prompt_toolkit, so this might be something I'm not able to fix. Will hopefully look into it in the coming few days and see what I can do.

**Comment by DismissedGuy** [comment]
Sorry for taking so long; I've been unable to reproduce this issue on my end. Does it also happen when you try to run it from source?

You could also try [this](https://github.com/RiiConnect24/RiiTag-RPC/actions/runs/5316140499) build, I doubt it will fix the issue but it includes an updated dependency (which should be completely unrelated but who knows..)

**Comment by TheAlienDrew** [comment]
@DismissedGuy, here's what my sleuthing was able to find:
```bash
# Commit 0276d65 = v1.1.3 = (version bump)
# Commit 9216ffd = ------ = (no exit or freeze bugs)
# Commit e0f4b73 = ------ = Exit bug first introduced, this was the sentry_sdk downgrade to 0.16.1
# Commit 83bb3b1 = v1.1.4 = (version bump)
# Commit c55962a = ------ = (no change)
# Commit 5d96428 = v1.1.5 = (version bump)
# .... skip ....            (no changes)
# Commit 6536b99 = v1.2.0 = (version bump)
# Commit cf1d2b5 = ------ = Completely broken, but probably adds to freeze bug???
# Commit 5551658 = ------ = Freeze bug first introduced
# .... skip ....            (no changes)
# Commit 5551658 = v1.3.0 = (version bump)
# .... skip ....            (no changes)
# Commit 4b646fb = v1.3.1 = (version bump)
# .... skip ....            (no changes)
# Commit 17b602b = ------ = Works around exit bug, freeze bug still exists
```

**Comment by TheAlienDrew** [comment]
From what I gather on 5551658, I think it might be the "clickable buttons" that are causing the issue with the freezing on the Pi.

I know for sure (at least when it comes to the GUI libraries), that the Pi usually doesn't have support for all features, when compared to being ran on Windows/Mac/Linux (non-aarch64).

**Comment by TheAlienDrew** [comment]
When I have the time, I'll go through the commits made in https://github.com/RiiConnect24/RiiTag-RPC/pull/25, and see at what point the freeze bug first appears (unless you get to that before me).

**Comment by DismissedGuy** [comment]
Woah, thanks a lot for the extensive research! I think your intuition might be correct; it looks like ArmCord uses [arRPC](https://arrpc.openasar.dev/), which most likely does not play nicely with pypresence. Their website does mention that parties are unsupported, maybe buttons fall under the same category? It might also be the fact that we're now sending full image URLs instead of asset names.

I think the best way to fix this would be to somehow detect the raspberry pi platform and disable the offending feature in that case. I'll look into it more tomorrow.

**Comment by TheAlienDrew** [comment]
Well this is interesting, up to even the last commit in that branch, #25, doesn't exhibit the freeze bug.

Only after it's been merged into the master does it start to exhibit the bug...

I don't understand why that would be unless it was a much earlier change before the merge (not causing any issues yet) bashing with a change in the merge.

**Comment by TheAlienDrew** [comment]
> Woah, thanks a lot for the extensive research! I think your intuition might be correct; it looks like ArmCord uses [arRPC](https://arrpc.openasar.dev/), which most likely does not play nicely with pypresence. Their website does mention that parties are unsupported, maybe buttons fall under the same category? It might also be the fact that we're now sending full image URLs instead of asset names.
> 
> I think the best way to fix this would be to somehow detect the raspberry pi platform and disable the offending feature in that case. I'll look into it more tomorrow.

Doesn't seem to be the buttons, but yeah, it does seem likely that pypresence could be causing issues with arRPC, unless @libewa also experiences the same bug (asking in https://github.com/RiiConnect24/RiiTag-RPC/issues/31#issuecomment-1616268837)

**Comment by TheAlienDrew** [comment]
@DismissedGuy Huh I guess it really must be arRPC, because testing the emulated x64 Discord client doesn't exhibit either of the bugs that I found while testing with ArmCord.

**Comment by TheAlienDrew** [comment]
@DismissedGuy
Here's some handy code you'll be needing to detect and disable some features if ArmCord is running:
```python3
import psutil
"armcord" in (p.name() for p in psutil.process_iter(attrs=['name']))
```

**Comment by DismissedGuy** [comment]
> Well this is interesting, up to even the last commit in that branch, #25, doesn't exhibit the freeze bug.
> 
> Only after it's been merged into the master does it start to exhibit the bug...
> 
> I don't understand why that would be unless it was a much earlier change before the merge (not causing any issues yet) bashing with a change in the merge.

That's... interesting, indeed. So just to summarize, the bug doesn't exist in 6536b99, it doesn't exist in cf1d2b5 either, but it does in 5551658?

That's weird to me because the second is a merge from master into rpc-improvements, and the third is a merge from rpc-improvements back into master. According to the network graph there are no commits to either master or rpc-improvements between the 2nd and 3rd commit. I do remember solving a merge conflict for that branch, but that was related to [this](https://github.com/RiiConnect24/RiiTag-RPC/commit/5551658098efc874c0ef5388f83425d26602a5f9#diff-d826d8f30d78e1d251e86720a5d3574829d22d05587c82be9338e80414fbc959R7) line, and I really don't see how that would cause it...

I was thinking it could be related to 5c16cc9, which was made directly to master after rpc-improvements was merged. But that wouldn't make sense if the bug exists in 5551658, since it was made after that.

**Comment by TheAlienDrew** [comment]
> > Well this is interesting, up to even the last commit in that branch, #25, doesn't exhibit the freeze bug.
> > Only after it's been merged into the master does it start to exhibit the bug...
> > I don't understand why that would be unless it was a much earlier change before the merge (not causing any issues yet) bashing with a change in the merge.
> 
> That's... interesting, indeed. So just to summarize, the bug doesn't exist in [6536b99](https://github.com/RiiConnect24/RiiTag-RPC/commit/6536b994c50adbcea6230870a38b37e54e6ea931), it doesn't exist in [cf1d2b5](https://github.com/RiiConnect24/RiiTag-RPC/commit/cf1d2b575beefc38955e7f66f861c8b2d9b9dec6) either, but it does in [5551658](https://github.com/RiiConnect24/RiiTag-RPC/commit/5551658098efc874c0ef5388f83425d26602a5f9)?
> 
> That's weird to me because the second is a merge from master into rpc-improvements, and the third is a merge from rpc-improvements back into master. According to the network graph there are no commits to either master or rpc-improvements between the 2nd and 3rd commit. I do remember solving a merge conflict for that branch, but that was related to [this](https://github.com/RiiConnect24/RiiTag-RPC/commit/5551658098efc874c0ef5388f83425d26602a5f9#diff-d826d8f30d78e1d251e86720a5d3574829d22d05587c82be9338e80414fbc959R7) line, and I really don't see how that would cause it...
> 
> I was thinking it could be related to [5c16cc9](https://github.com/RiiConnect24/RiiTag-RPC/commit/5c16cc978bc47602c02a242d6f43ab3a806567c6), which was made directly to master after rpc-improvements was merged. But that wouldn't make sense if the bug exists in [5551658](https://github.com/RiiConnect24/RiiTag-RPC/commit/5551658098efc874c0ef5388f83425d26602a5f9), since it was made after that.

Yep, I was thinking the same thing. Bug probably introduced from the double merge, aka, a combination of two different changes that on their own don't cause the issue.

But alas, at least we know that it's just something that affects arRPC, so it's probably just that they're missing some feature support I'd assume.

**Comment by TheAlienDrew** [comment]
Which, I've already left them an issue for them to check out, if they ever will: https://github.com/OpenAsar/arrpc/issues/38

**Comment by DismissedGuy** [comment]
I will try to create a very minimal example tomorrow that uses pypresence to set the same fields as riitag-rpc does. That should help narrow down which exact fields are causing the problem.

**Comment by DismissedGuy** [comment]
Whoops, totally didn't forget about this. Here's a minimal example:

```py
import logging
import time

import pypresence

OPTIONS = {
    'details': f'Playing Some Game',
    'state': f'Playing on Wii U',
    'start': 1,

    'large_image': 'https://art.gametdb.com/wii/cover/US/HCFE.png',
    'large_text': 'Some Game',

    'small_image': 'logo',
    'small_text': 'tag.rc24.xyz',

    'buttons': [
        {'label': 'View RiiTag', 'url': f'https://tag.rc24.xyz/user/311869975579066371'}
    ]
}

logging.basicConfig(level=logging.DEBUG)


logging.info('Creating presence object')
pres = pypresence.Presence(client_id='749633517813628968')

logging.info('Connecting')
pres.connect()

logging.info('Setting presence')
pres.update(**OPTIONS)

logging.info('Sleeping for 10 seconds...')
time.sleep(10)

logging.info('Clearing presence')
pres.clear()
```

Can you try that and see if it has the same issue?

**Comment by TheAlienDrew** [comment]
> Whoops, totally didn't forget about this. Here's a minimal example:
> 
> ```python
> import logging
> import time
> 
> import pypresence
> 
> OPTIONS = {
>     'details': f'Playing Some Game',
>     'state': f'Playing on Wii U',
>     'start': 1,
> 
>     'large_image': 'https://art.gametdb.com/wii/cover/US/HCFE.png',
>     'large_text': 'Some Game',
> 
>     'small_image': 'logo',
>     'small_text': 'tag.rc24.xyz',
> 
>     'buttons': [
>         {'label': 'View RiiTag', 'url': f'https://tag.rc24.xyz/user/311869975579066371'}
>     ]
> }
> 
> logging.basicConfig(level=logging.DEBUG)
> 
> 
> logging.info('Creating presence object')
> pres = pypresence.Presence(client_id='749633517813628968')
> 
> logging.info('Connecting')
> pres.connect()
> 
> logging.info('Setting presence')
> pres.update(**OPTIONS)
> 
> logging.info('Sleeping for 10 seconds...')
> time.sleep(10)
> 
> logging.info('Clearing presence')
> pres.clear()
> ```
> 
> Can you try that and see if it has the same issue?

That appears to work fine.

**Comment by TheAlienDrew** [comment]
Ah, but as for ArmCord, stuck on clearing presence.

**Comment by TheAlienDrew** [comment]
However, I do notice that it actually clears the presence, just hangs in the script in the terminal window.

**Comment by DismissedGuy** [comment]
Ah, that explains a lot. I just read through pypresence's source code and apparently there are some hidden options to lower the socket timeouts. I've set those to 5 seconds in the [latest build](https://github.com/RiiConnect24/RiiTag-RPC/actions/runs/5524431900), which should hopefully remedy the issue.

Note that the freezing issue will most likely still be there, but I'm hoping it will start working after those 5 seconds. Nothing I can do about that unfortunately (other than lowering them even further, but I'd rather not to be honest)

**Comment by TheAlienDrew** [comment]
After the 5 seconds, it still appears to be frozen in console, but I guess we'll just have to wait to let OpenAsar figure that one out over at https://github.com/OpenAsar/arrpc/issues/38

**Comment by DismissedGuy** [comment]
Unfortunate. I might be able to debug the exact cause if the problem also exists when running arrpc on my device, will let you know if I know more.

**Comment by DismissedGuy** [comment]
I'm able to repro this issue when running arRPC locally. Haven't found the exact cause yet, but I did notice that pypresence was outdated so those extra arguments I added didn't actually do anything. A new build should pop up under the actions tab in a few minutes, that should actually enfore that 5 second timeout.

**Comment by DismissedGuy** [comment]
Found and fixed the bug in https://github.com/OpenAsar/arrpc/pull/39, so this issue should be fixed when (if) that gets merged and adopted into armcord. I'm going to close this issue for now; if it still persists with this fix feel free to reopen.
